DOCSP-30509: Usage examples landing page (#61)
mongodb/docs-rust: source/usage-examples.txt

.. _rust-usage-examples:

==============
Usage Examples
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Overview
--------

Usage examples provide convenient starting points for popular MongoDB
operations. Each example provides the following:

> - A full asynchronous and synchronous Rust program that you can run in
>   your own environment

Reviewer: ```suggestion
- A Rust program that you can run in your own environment
```

- The expected output of the program

> 

Reviewer: Add the info about sync async here:
```suggestion

Every usage example provides the same example using both the asynchronous and synchronous runtimes.
.. To learn more about runtimnes... link to future page
```

How to Use the Usage Examples
-----------------------------

These examples use the :atlas:`sample datasets </sample-data>` provided by
Atlas. Each example runs one or more operations on the ``restaurants``
collection in the ``sample_restaurants`` database. You can load the sample
datasets into your database on the free tier of MongoDB Atlas by following the
:atlas:`Get Started with Atlas Guide </getting-started/#atlas-getting-started>`
or you can :guides:`import the sample dataset into a local MongoDB instance
</server/import/>`.

> Once you import the datasets, you can copy and paste a usage example into your
> development environment of choice. Follow the :ref:`rust-quickstart` to learn
> more about getting started with the {+driver-short+}.

Reviewer: ```suggestion
Once you import the datasets, you can copy and paste a usage example into your
development environment of choice. Follow the :ref:`rust-quickstart` to learn
more about creating a sample app that uses the {+driver-short+}.
```

Connection String
~~~~~~~~~~~~~~~~~

> After you have copied a usage example into your development environment, you

Reviewer: ```suggestion
After you copy a usage example into your development environment, you
```

must edit the connection URI to connect the example to your MongoDB instance:

.. code-block:: none
   :copyable: false

>    // Replace the bracketed text with your MongoDB deployment's connection string

Reviewer: ```suggestion
   // Replace the placeholder with your MongoDB deployment's connection string
```

.. _rust-usage-examples:

==============
Usage Examples
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Overview
--------

Usage examples provide convenient starting points for popular MongoDB
operations. Each example provides the following:

- A Rust program that you can run in your own environment
- The expected output of the program

Each usage example provides the same example using both the asynchronous

> and synchronous runtimes.

Author: I included information about the runtimes in the list itself, but kept the sentence directing readers to further information on runtimes after the list (commented out in a TODO for now)

.. TODO: link to the runtimes guide

> How to Use the Usage Examples

Reviewer 2: Suggestion:
Since this starts with "How to...", I think it might be better to represent as a [task](https://www.mongodb.com/docs/meta/style-guide/style/tasks/#std-label-tasks), and the information on this page should be converted to fit that model. This can be ticketed out if it seems out of scope.

One alternative that I can think of is to remove details on how to perform the tasks and just provide links to the different options the reader can follow for setup. E.g.

```
Run the Usage Examples
---

To run the usage examples, load the Atlas sample datasets into your MongoDB deployment. To learn how to load the datasets into your Atlas cluster, see ... To learn how to load the datasets into your local deployment, see...

Each example contains placeholders including one for the connection string. To learn how to find your connection string, see...
```

Author: Reformatted this section to fit the task model

-----------------------------

These examples use the :atlas:`sample datasets </sample-data>` provided by
Atlas. Each example runs one or more operations on the ``restaurants``

> collection in the ``sample_restaurants`` database. You can load the sample

Author: Since this info is now included in a paragraph introducing the Atlas sample data setup steps, I think it's relevant -- let me know what you think

datasets into your database on the free tier of MongoDB Atlas by following the
:atlas:`Get Started with Atlas Guide </getting-started/#atlas-getting-started>`
or you can :guides:`import the sample dataset into a local MongoDB instance

> </server/import/>`.

Reviewer 2: Issue:

I noticed that this link redirects to atlas/sample-data which only shows how to import the sample dataset into Atlas and not a local instance. I wasn't able to find a good page to link to in the five minutes I spent looking, so maybe there's some reason why it's not in the docs. Could you reach out to the other docs teams in our Slack channels for advice on a replacement url?

Author: Since a page on this topic doesn't exist (according to JW's response in #docs-writers), I just took this part out

Once you import the datasets, you can copy and paste a usage example into your
development environment of choice. Follow the :ref:`rust-quickstart` to learn

> more about creating an application that uses the {+driver-short+}.

Reviewer 2: Suggestion:

No change needed, but if you wanted to improve this description, a couple areas that I think can be addressed are:

- Provide information that explains what "creating an application" has to do with "copy and paste a usage example". I think there's some assumptions the reader needs to make to understand this, so it could be beneficial to provide some explanation.
- Use the imperative mood to explain more concisely and definitively how the reader can get to the goal.

Author: I reworded to the imperative mood, and also to clarify that you're pasting into an application file

.. _rust-usage-examples:

==============
Usage Examples
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Overview
--------

> Usage examples enable you to set up, run, and gain a broad understanding of
> a MongoDB operation. Each usage example includes the following:

Reviewer: ```suggestion
Usage examples enable you gain a broad understanding of common MongoDB operations by 
demonstrating how to organize and run a small sample application. Each usage example includes the following:
```

example includes the following:

> - A general overview of the MongoDB operation

Reviewer 2: Suggestion:

Per the style guide on [List Items](https://www.mongodb.com/docs/meta/style-guide/style/lists/#list-items):
> Avoid using articles (a, an, the) to start list items.

Also, I think "overview" can be too indefinite and is usually interpreted as a "big picture" description which probably doesn't match what we intend to include. I would recommend something like this:

"Description of the MongoDB operation"

> - One asynchronous and one synchronous Rust program that you can run
>   in your own environment

Reviewer: ```suggestion
- Asynchronous and synchronous Rust code examples that you can run
  in your own environment
```

> - The expected output of the program

Reviewer: ```suggestion
- The expected output of the code example
```

> - The expected output of the code example

Reviewer 2: ```suggestion
- Output printed by the code example
```

.. TODO: To learn more about the runtimes offered by the {+driver-short+},
   see <add link to runtimes page>.

How to Use the Usage Examples
-----------------------------

The usage examples run operations on the ``restaurants`` collection in the

> ``sample_restaurants`` database. This dataset is one of the :atlas:`sample datasets
> </sample-data>` provided by Atlas. Before running the programs included in the

Reviewer: S: don't link here, since the instructions on loading the sample data are included in the procedure step "Retrieve the Sample Datasets" anyways

```suggestion
``sample_restaurants`` database. This dataset is one of the sample datasets
included in the Atlas sample data. Before running the programs included in the
```

> The usage examples run operations on the ``restaurants`` collection in the
> ``sample_restaurants`` database. This dataset is one of the sample datasets
> provided in the Atlas sample data. Before running the programs included in
> the usage examples, perform the following actions:

Reviewer 2: Suggestions:

- I think presenting the usage examples on data would be helpful. 
- I think the reason why they should perform the following actions is missing.
- I think the option not to perform the setup should be mentioned since the usage examples will still run, but probably not match the output.
- I think it's not important to mention which database or collection since the instructions only show how to load all the sample data at once and not individually.
 
```suggestion
The usage examples are designed to run operations on a MongoDB deployment that
contains the Atlas sample datasets. When you run the example code without the Atlas
data, the output may not match. To set up your MongoDB deployment with the Atlas data
and run the example code in your development environment, perform the following actions:
```

> usage examples, perform the following actions to access the sample datasets in
> your development environment.

Reviewer: S: edit this sentence. Right now, this sentence only describes step 1

```suggestion
usage examples, perform the following actions:
```

.. procedure::
   :style: connected

>    .. step:: Retrieve the Sample Datasets

Reviewer 2: Suggestion:
I think "Load" is a more accurate verb and is what is used in the Atlas docs in the title "Load Sample Data". I think "Retrieve" used in this context only describes the process of downloading them.

      
      Follow the :atlas:`Get Started with Atlas Guide </getting-started/#atlas-getting-started>`

>       to import the sample datasets into your database on the free tier of MongoDB Atlas.

Reviewer: S: link to this page instead of the Getting started guide: https://www.mongodb.com/docs/atlas/sample-data/

Alternatively, you should provide more information in this section that describes that the user needs to also have an atlas account before even loading the sample data

Author: Since the current link (Getting Started guide) includes steps to create an account, create/connect to a cluster, and load sample data, would it be better to leave it instead of just linking to the Load Sample Data page? I think that way I can leave out info about making an account, but let me know what you think

>    .. step:: Use the Sample Datasets in your Development Environment

Reviewer: S: update this step name to more accurately match the step text

```suggestion
   .. step:: Add the Usage Example Code to your Development Environment
```

>       Copy the usage example program that you wish to run. Then, navigate to a file in

Reviewer 2: Issue:

From the style guide [terms](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#w):
> Use want, or another appropriate word, instead of desire or wish.

Suggestion:

In addition to following the style guide guidance, I think it would be helpful to include instructions from scratch in which no pre-existing file is assumed.

I think the following could be a good replacement for these two sentences:
"Copy the example code from the Usage Example page and paste it into a new file in your preferred directory."

      your application and paste the program.
      
      To learn more about creating an application that uses the {+driver-short+}, follow the
      :ref:`rust-quickstart`.

>    .. step:: Edit the Connection String

Reviewer 2: Suggestion:
I think "Edit" might not be the most precise verb and "Connection String" might not portray the object since it's a placeholder. I think "Replace the Connection String Placeholder" would match the wording in Rea's suggestion and be easier to understand.

>       In your application file, edit the connection URI to connect to your MongoDB instance:

Reviewer: ```suggestion
      In your application file, replace the connection string placeholder so that the program connects to your 
      MongoDB deployment:
```

      .. code-block:: none
         :copyable: false

>          // Replace the placeholder with your MongoDB deployment's connection string

Reviewer: S: if you accept my previous comment, this code comment is no longer needed

>       In your application file, replace the connection string placeholder so that the program
>       connects to your MongoDB deployment:
> 
>       .. code-block:: rust
>          :copyable: false
> 
>          let uri = "<connection string>";

Reviewer 2: Suggestion:

I think the code example may not demonstrate what the user needs to do. 

```suggestion
In your example code, replace the ``connection string`` placeholder with your MongoDB deployment connection string. For example, if your connection string is ..., your connection string assignment should resemble the following:
...
```

After completing these steps, you can run a program in your development environment
and see the output described in the **Expected Output** section of the corresponding

> usage example.

Reviewer 2: Suggestion:

I think this isn't the completion state for "how to use a usage example" since they haven't run the program yet, and needs to continue to use the imperative mood ("do this") rather than one of possibility ("you can").

I think running the code should be a separate step even though it might need to be generalized for different development environments.

"Run your application in your development environment. If you run your Rust applications on the command line, the command should resemble the following: ..."

> After completing these steps, you can run a program in your development environment
> and see the output described on the program's corresponding usage example page.

Reviewer: ```suggestion
After completing these steps, you can run a program in your development environment
and see the output described in the **Expected Output** section of the corresponding usage example.
```

> 

Reviewer: S: delete extra line

.. Available Usage Examples
.. ------------------------

.. TODO: add Usage Example pages as they are created

>     - :ref:`Find a Document<rust-find-one-usage>`
>     - :ref:`Find Multiple Documents<rust-find-usage>`
>     - :ref:`Insert a Document<rust-insert-one-usage>`
>     - :ref:`Insert Multiple Documents<rust-insert-many-usage>`
>     - :ref:`Update a Document<rust-update-one-usage>`
>     - :ref:`Update Multiple Documents<rust-update-many-usage>`
>     - :ref:`Replace a Document<rust-replace-usage>`
>     - :ref:`Delete a Document<rust-delete-one-usage>`
>     - :ref:`Delete Multiple Documents<rust-delete-many-usage>`
>     - :ref:`Count Documents<rust-count-usage>`
>     - :ref:`Find Distinct Values<rust-distinct-usage>`

Reviewer: S: though we will edit these as we work on the tickets, add the space before the <
```suggestion
    - :ref:`Find a Document <rust-find-one-usage>`
    - :ref:`Find Multiple Documents <rust-find-usage>`
    - :ref:`Insert a Document<rust-insert-one-usage>`
    - :ref:`Insert Multiple Documents<rust-insert-many-usage>`
    - :ref:`Update a Document<rust-update-one-usage>`
    - :ref:`Update Multiple Documents<rust-update-many-usage>`
    - :ref:`Replace a Document<rust-replace-usage>`
    - :ref:`Delete a Document<rust-delete-one-usage>`
    - :ref:`Delete Multiple Documents<rust-delete-many-usage>`
    - :ref:`Count Documents<rust-count-usage>`
    - :ref:`Find Distinct Values<rust-distinct-usage>`
```

.. _rust-usage-examples:

==============
Usage Examples
==============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 1
   :class: singlecol

Overview
--------

Usage examples show runnable code examples to demonstrate frequently used MongoDB
operations. Each usage example includes the following:

- Description of the MongoDB operation
- Asynchronous and synchronous Rust code examples that you can run in
  your own environment
- Output printed by the code example

.. TODO: To learn more about the runtimes offered by the {+driver-short+},
   see <add link to runtimes page>.

How to Use the Usage Examples
-----------------------------

The usage examples are designed to run operations on a MongoDB deployment that contains

> the Atlas sample datasets. When you run the example code without the Atlas data, the output

Reviewer 2: I think "sample" could be a slightly more precise adjective than "Atlas" in this case.

```suggestion
the Atlas sample datasets. When you run the example code without the sample data, the output
```

may not match. 

> Follow this tutorial to set up your MongoDB deployment with the Atlas data and run the

Reviewer 2: Suggestion: 
Similar to the previous comment:

```suggestion
Follow this tutorial to set up your MongoDB deployment with the sample data and run the
```

example code in your development environment. Before performing the following actions,
ensure that you create an Atlas account and deploy a cluster. For information about setting
up an account and a cluster, see the :atlas:`Get Started with Atlas Guide
</getting-started/#atlas-getting-started>`.

.. procedure::
   :style: connected

   .. step:: Load the Sample Datasets
      
      Follow the instructions on the :atlas:`Load Sample Data </sample-data>` page
      to load the sample datasets into your database deployment.

   .. step:: Add the Usage Example Code to Your Development Environment

      Copy the example code from the usage example page and paste it into a new file in

>       your preferred application directory.

Reviewer 2: Suggestion:

I think "application directory" could be confused with a system directory that holds applications such as "Applications" in MacOS, and omitting "application" could lower the chance of confusion.

```suggestion
      your preferred directory.
```

      
      To learn more about creating an application that uses the {+driver-short+}, follow the
      :ref:`rust-quickstart`.

   .. step:: Replace the Connection String Placeholder

      In your example code, replace the ``connection string`` placeholder with your MongoDB
      deployment connection string. For example, if your connection string is

>       ``"mongodb://localhost:27017"``, your connection string assignment should resemble the

Reviewer 2: Suggestion:

Since this section mentions Atlas deployments, I think it would be better to use the DNS seedlist format.
Also, I think you can replace the subjunctive "should" as follows:

```suggestion
      ``"mongodb+srv://mongodb-example:27017"``, your connection string assignment resembles the
```
